rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,962 @@
+
+    `value` will always be evaluated on all replicas.
+
+    Example with two replicas:
+      Replica 0:: `value`: {'a': 1, 'b': [40,  1]}, `source_replica_id`: 1
+      Replica 1:: `value`: {'a': 3, 'b': [ 2, 98]}, `source_replica_id`: 1
+
+      Replica 0:: result: {'a': 3, 'b': [ 2, 98]}
+      Replica 1:: result: {'a': 3, 'b': [ 2, 98]}
+
+    Args:
+      value: The nested structure of `Tensor`s to broadcast.
+        The structure must be compatible with `tf.nest`.
+      source_replica_id: The ID of the replica from which to broadcast.
+
+    Returns:
+      A `Tensor` nest with the `value` from the source replica.
+    """
+```
+
+#### InputReplicationMode
+
+```python
+class InputReplicationMode(enum.Enum):
+  SINGLE = 0
+  PER_WORKER = 1
+  PER_REPLICA = 2
+```
+
+##### SINGLE
+
+The input function will be called once, typically on the first worker. The
+entire input pipeline ops then live on that worker. All replicas (on that and
+other workers) will dequeue from a single `Dataset` created on that worker.
+
+This mode is not supported by between-graph Distribution Strategy
+implementations.
+
+##### PER_WORKER
+
+The input function will be called on each worker independently, creating as many
+input pipelines as number of workers. Replicas will dequeue from the local
+`Dataset` on their worker. Replicator doesn't manage any state sharing between
+such separate input pipelines.
+
+##### PER_REPLICA
+
+The input function will be called for every replica, on the corresponding
+worker, thus creating as many input pipelines as number of replicas. Each
+replica will dequeue from its own `Dataset`.
+
+#### InputContext
+
+An instance of `InputContext` is passed to the input function.
+
+```python
+class InputContext(object):
+  @property
+  def num_replicas(self) -> int:
+    """Returns the total number of compute replicas."""
+
+  @property
+  def num_replicas_in_sync(self) -> int:
+    """Returns the number of compute replicas training in sync."""
+
+  @property
+  def input_pipeline_id(self) -> int:
+    """Returns the input pipeline ID.
+
+    If `replication_mode` == `SINGLE`, this is always `0`.
+    If `replication_mode` == `PER_WORKER`, this is the worker ID.
+    If `replication_mode` == `PER_REPLICA`, this is the compute replica ID.
+    """
+
+  @property
+  def num_input_pipelines(self) -> int:
+    """Returns the number of input pipelines.
+
+    If `replication_mode` == `SINGLE`, this is always `1`.
+    If `replication_mode` == `PER_WORKER`, this is the number of workers.
+    If `replication_mode` == `PER_REPLICA`, this is `num_replicas`.
+    """
+```
+
+#### InputIterator
+
+An instance of `InputIterator` is returned by `prepare_input` and can be passed
+to `run`.
+
+```python
+class InputIterator(Generic[T]):
+  """An input iterator, intended to be passed to `Replicator.run`."""
+
+  def get_next(self) -> List[T]:
+    """Returns the next inputs for each replica."""
+
+  def reinitialize(self) -> List[tf.Operation]:
```

Comment:
Can you clarify what reinitialization is intended to accomplish? When is it necessary?

Reply:
Yes, it's actually the same as calling `initialize` again on an initializable iterator. In the upcoming udpate, we are changing this to be called just `initialize` so it's more clear and matches the regular tf.data iterator methods. An example of when it's useful to initialize the iterator again is during evaluation (See the evaluation example). Since the user is managing the training loop themselves, this is useful for them.